Break up ParseFrontendOptions and reorganize module and output files computations.

davidungar opened the pull request:
<!-- What's in this pull request? -->
Break up ParseFrontendOptions and reorganize module and output files computations.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->


<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

davidungar commented:
@swift-ci please smoke test os x platform

davidungar commented:
@swift-ci please smoke test os x platform

davidungar commented:
@swift-ci please smoke test os x platform

davidungar commented:
@swift-ci please smoke test os x platform

davidungar commented:
@swift-ci please smoke test os x platform

davidungar commented:
@swift-ci please smoke test os x platform

davidungar commented:
@swift-ci please smoke test linux platform

jrose-apple commented:
Sorry, David, mind rebasing this? There are a ton of unrelated changes that make it hard to review all at once, and then the piecemeal fixes make it hard to review commit-by-commit.

davidungar commented:
@swift-ci please smoke test

davidungar commented:
Jordan, I have rebased. Should be easier to look at, now.

davidungar commented:
@swift-ci please smoke test

davidungar commented:
@swift-ci please smoke test linux platform

davidungar commented:
@swift-ci please smoke test

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -222,26 +222,161 @@ class ArgsToFrontendInputsConverter {
     return false;
   }
+};
+    class FrontendArgsToOptionsConverter {
```
What happened to the indentation here?

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -222,26 +222,161 @@ class ArgsToFrontendInputsConverter {
+
+  // This is a separate function so that it shows up in stack traces.
+  LLVM_ATTRIBUTE_NOINLINE
+  static void debugFailWithCrash() { LLVM_BUILTIN_TRAP; }
```
Again, these two functions should remain top-level, not members.

davidungar replied:
Moved them. But there's something here I don't understand.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -222,26 +222,161 @@ class ArgsToFrontendInputsConverter {
+  LLVM_ATTRIBUTE_NOINLINE
+  static void debugFailWithCrash() { LLVM_BUILTIN_TRAP; }
+
+  void computeDebugCrashGroup();
```
Nothing is being "computed" here, though. `handle` would work.

jrose-apple replied:
I actually like `handle` for the various option-reading functions too, though that's more of a taste issue instead of being actively misleading. Does it make sense how these are different from `computeModuleName`?

davidungar replied:
It is setting the CrashMode option. These functions all set something, and I think I may have originally called them set<something>. The value that is set is the result of a (possibly trivial) computation, as I tend to frame it. But, I'm happy to name them whatever you like. My preference would be a consistent prefix.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -222,26 +222,161 @@ class ArgsToFrontendInputsConverter {
+  void computePrintStatsOptions();
+  void computeTBDOptions();
+  void setUnsignedIntegerArgument(options::ID optionID, unsigned max,
+                                  unsigned &valueToSet);
```
Nitpick: weird to declare this in the middle of all the more high-level functions.

davidungar replied:
It's used in the same context. Do you still want it moved?

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -222,26 +222,161 @@ class ArgsToFrontendInputsConverter {
+  bool computeFallbackModuleName();
+  bool computeOutputFilenames();
+  bool deriveOutputFilenameFromInputFile();
+  bool deriveOutputFilenameForDirectory(llvm::StringRef outputDir);
```
Nitpick: No `llvm` on StringRef.

davidungar replied:
Fixed

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -222,26 +222,161 @@ class ArgsToFrontendInputsConverter {
+  /// Returns the output filenames on the command line or in the output
+  /// filelist. If there was an error (reading the list) returns None. If there
+  /// were neither -o's nor an output filelist, returns an empty vector.
+  const llvm::Optional<const std::vector<std::string>> &
```
No need to make the template parameter type `const` here. It's already a value.

Also, nitpick: no `llvm` on Optional.

davidungar replied:
Fixed.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -222,26 +222,161 @@ class ArgsToFrontendInputsConverter {
+  /// were neither -o's nor an output filelist, returns an empty vector.
+  const llvm::Optional<const std::vector<std::string>> &
+  getOutputFilenamesFromCommandLineOrFilelist();
+  bool hasAnUnusedOutputPath() const;
```
Nitpick: `An` is superfluous here.

jrose-apple replied:
Also unhappy that a function named like a predicate also emits diagnostics. Maybe `checkForUnusedOutputPaths`?

davidungar replied:
Fixed.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -222,26 +222,161 @@ class ArgsToFrontendInputsConverter {
+  void computeImportObjCHeaderOptions();
+  void computeImplicitImportModuleNames();
+  void computeLLVMArgs();
+  const std::vector<std::string>
```
`const` has no effect on return types in C++.

davidungar replied:
Fixed, thanks.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -222,26 +222,161 @@ class ArgsToFrontendInputsConverter {
+  void computeImplicitImportModuleNames();
+  void computeLLVMArgs();
+  const std::vector<std::string>
+  readOutputFileList(const StringRef filelistPath) const;
```
…and isn't really useful on by-value arguments either. This one's a bit of a conflict because it *does* mean something in the *body* of the function, but it's distracting when reading the function signature.

davidungar replied:
I would like to push back. It guarantees that the body cannot contain "filelistPath = <foo>".
In this case, it's not a big deal because the function body is short, but I would like to cultivate the habit.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -365,115 +475,109 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
-    Diags.diagnose(SourceLoc(), diag::error_immediate_mode_primary_file);
-    return true;
-  }
+    return Opts.RequestedAction; // no change
```
Does this case come up in practice? It seems weird that this method reads from `Opts` at all.

jrose-apple replied:
Maybe it should return `FrontendOptions::ActionType::NoneAction`?

davidungar replied:
I agree, but was reluctant to make that change. I'll run it up the flagpole, and see which test doesn't salute.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -499,184 +602,221 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
+  Diags.diagnose(SourceLoc(), DID, Opts.ModuleName, A == nullptr);
+  Opts.ModuleName = "__bad__";
+  return false; // Must continue to run to pass test:
+                // multifile.protocol-conformance.swift
```
Again, please do not mention tests by name in the source. Tests are more ephemeral than the source, and we should figure out what's actually going on here. Most likely we should just change the test.

davidungar replied:
Test name has been removed.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -499,184 +602,221 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
+    Opts.ModuleName = "REPL";
+    return false;
+  }
+  // In order to pass Driver/options.swift test must leave ModuleName empty
```
Here too.

davidungar replied:
Test name has been removed.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -499,184 +602,221 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
+    // The first test is: RUN: not %swift 2>&1 | %FileCheck %s
+    // -check-prefix=CHECK1 Which invokes swift with no input arguments and
+    // expects the following error: "no frontend action was selected"
+    return false;
```
I think this means we should just bail out earlier, where "no frontend action was selected", and not make it this far. But that can be changed later.

davidungar replied:
Added a comment to that effect.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -499,184 +602,221 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
+    // expects the following error: "no frontend action was selected"
+    return false;
+  }
+  const llvm::Optional<const std::vector<std::string>> &outputFilenames =
```
Nitpick: no `llvm` on Optional.

davidungar replied:
Fixed.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -499,184 +602,221 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
-    case FrontendOptions::ActionType::EmitSIB:
-      Suffix = SIB_EXTENSION;
-      break;
+  bool isOutputAUniqueOrdinaryFile =
```
Is it possible to make use of `getSingleOutputFilename` here?

davidungar replied:
I don't understand. The predicates differ, == 1, vs >= 1.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -499,184 +602,221 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
-      // These modes have no frontend-generated output.
-      Opts.OutputFilenames.clear();
-      break;
+  const llvm::Optional<const std::vector<std::string>>
```
Nitpick: No `llvm` on Optional.

davidungar replied:
Fixed.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -499,184 +602,221 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
-      break;
+  // Filelist names do not get altered.
+  if (outputFilenamesFromCommandLineOrFilelist->empty()) {
+    // Frontend was invoked directly, without driver.
```
This comment is unclear. It's true that this only happens when the frontend is invoked directly, and that we don't care so much about it in the driver case, but this shouldn't be a *signifier* of that.

davidungar replied:
Sorry, I need clarification here. Do you want the comment reworded to say:
"This case only happens when the frontend is invoked directly, without the driver."

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -499,184 +602,221 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
+  if (outputFilenamesFromCommandLineOrFilelist->size() == 1) {
+    StringRef outputFilename = (*outputFilenamesFromCommandLineOrFilelist)[0];
+    if (llvm::sys::fs::is_directory(outputFilename)) {
+      // Only used for testing & when invoking frontend directly.
```
This one is good. :-)

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -499,184 +602,221 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
     }
+    // Could be -primary-file (1), or -wmo (non-threaded w/ N files)
+    Opts.OutputFilenames = *outputFilenamesFromCommandLineOrFilelist;
+    return false;
```
Why not fall through?

davidungar replied:
I thought the same thing, but the differing comments swayed me the other way.
The fall-through is for a different situation, even though the code (today, but not tomorrow) is the same. What do you think?

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -499,184 +602,221 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
-      break;
-    }
+// No output filename was specified.
+// Determine the correct output filename.
```
Orphaned comments?

davidungar replied:
Reunited it with it's sub-comment.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -499,184 +602,221 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
+    return false;
+  }
+  std::string baseName = computeBaseNameOfOutput();
+  if (baseName == "") {
```
Prefer `baseName.empty()`.

davidungar replied:
Happy to fix it.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -499,184 +602,221 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
-      }
-    }
+// An output directory was specified.
+// Determine the correct output filename.
```
Orphaned comments?

davidungar replied:
Added an empty comment line after to connect to next comment.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -499,184 +602,221 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
+    StringRef outputDir) {
+
+  std::string baseName = computeBaseNameOfOutput();
+  if (baseName == "") {
```
`baseName.empty()` again.

davidungar replied:
Happy to fix it.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -499,184 +602,221 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
+    nameToStem = Opts.Inputs.primaryInputFilenameIfAny();
+  } else if (auto UserSpecifiedModuleName =
+                 Args.getLastArg(options::OPT_module_name)) {
+    nameToStem = std::string(UserSpecifiedModuleName->getValue());
```
There's an implicit conversion from `const char *` to `std::string`, so no need for the explicit constructor.

davidungar replied:
Thanks. Fixed!

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -697,175 +837,97 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
+    cachedOutputFilenamesFromCommandLineOrFilelist.emplace(
+        Args.getAllArgValues(options::OPT_o));
+  }
+  return cachedOutputFilenamesFromCommandLineOrFilelist;
```
This is never going to be empty in practice. Why not return a non-optional value?

davidungar replied:
My original idea was to signal an error opening the output filelist to the caller.
But that idea fell through the cracks.
But, since the compiler does not do that now, I have changed it to be non-optional.

jrose-apple commented on lib/Frontend/FrontendOptions.cpp:
```diff
@@ -57,7 +44,7 @@ bool FrontendInputs::shouldTreatAsSIL() const {
   }
   // If we have one primary input and it's a filename with extension "sil",
-  // treat the input as SIL.
+    // treat the input as SIL.
```
Stray tab here.

jrose-apple commented on lib/Frontend/FrontendOptions.cpp:
```diff
@@ -236,57 +195,276 @@ StringRef FrontendOptions::originalPath() const {
-    return StringRef();
+}
+
+bool FrontendOptions::hasUnusedDependenciesFilePath() const {
```
I still don't think this particular level of helpers is pulling its weight. What was your reasoning for not just expanding this `&&` into the caller?

davidungar replied:
I balked at repeating "Opts" three times. For instance:
if (!*Opts*.DependenciesFilePath.empty() &&
      !*Opts*.canActionEmitDependencies(*Opts.*RequestedAction))

But I'll make the change if you still prefer it.

jrose-apple replied:
No, I can't come up with a strong enough justification. Go ahead.

jrose-apple commented on lib/Frontend/FrontendOptions.cpp:
```diff
@@ -236,57 +195,276 @@ StringRef FrontendOptions::originalPath() const {
+}
+
+bool FrontendOptions::hasUnusedModuleDocOutputPath() const {
+  return !ModuleDocOutputPath.empty() && !canActionEmitModule(RequestedAction);
```
You've defined `canActionEmitModuleDoc` but you're not using it.

davidungar replied:
Fixed code to use it.

jrose-apple commented on lib/Frontend/FrontendOptions.cpp:
```diff
@@ -236,57 +195,276 @@ StringRef FrontendOptions::originalPath() const {
+  case ActionType::EmitSIL:
+  case ActionType::EmitAssembly:
+  case ActionType::EmitIR:
+    return true;
```
A number of these actually don't, so we might want to reinvestigate this later. But right now it's better to stay consistent with what we currently do.

davidungar replied:
Added comment.

jrose-apple reviewed (commented):
Many little comments but it's definitely converging now.

davidungar committed:
Break up ParseFrontendOptions and reorganize module and output files computations.

davidungar committed:
Incorperate advice from review.

Change “have” routines to “has”.
Use more consistent casing.
Remove spurious “DelayedFunctionParsing” option.
Move debugFail routines to top lexical level.
Rename and reorder declaration of functions in FrontendArgsToOptionsConverter.
Move, reword, and doxygenate comments for some of those functions.
Fix casing on some more setUp* functions.
Return NoneAction instead of existing RequestedAction in FrontendArgsToOptionsConverter::determineRequestedAction.
Remove test names and put in FIXME’s.
Remove “Jordan” from comments & reword.
Reorder if-then arms of FrontendArgsToOptionsConverter::computeOutputFilenames for readability.
Test for empty string instead of equality with “”.
Use hasUnusedModuleDocOutputPath.
Remove optionality from return type of getOutputFilenamesFromCommandLineOrFilelist.
Rename isPrimaryInputAFileAt to isThereAPrimaryInputWithAFilenameAt.
Added a FIXME in doesActionProduceOutput to reflect that some actions actually do not produce output.

davidungar commented:
@swift-ci please smoke test

jrose-apple commented on include/swift/Frontend/FrontendOptions.h:
```diff
@@ -162,15 +158,15 @@ class FrontendInputs {
   }
 
-  bool isPrimaryInputAFileAt(unsigned i) const {
+  bool isThereAPrimaryInputWithAFilenameAt(unsigned i) const {
```
This still doesn't make sense to me. Why is this using "a" at all?

davidungar replied:
Do you like "isInputPrimary" better? Can't recall exactly why it doesn't make sense.

jrose-apple replied:
That is better, yes. `isPrimaryInputAFileAt` is weird because it's asking two questions: "is the primary input a file" and "is the primary input at `i`", but "is the primary input a file at `i`" doesn't mean that. `isThereAPrimaryInputWithAFilenameAt` is better, but seems overly circuitous when you could just ask "is the input at `i` primary (and is it a file)".

It sounds like your *next* patch defines this whole thing away because (a) there are no non-file inputs, and (b) the inputs themselves know whether they are primary. So maybe it's not worth changing right now.

jrose-apple replied:
In general, our style omits "a"/"an"/"the", so if you're including it I want it to be significant.

jrose-apple commented on include/swift/Option/FrontendOptions.td:
```diff
@@ -21,6 +21,10 @@ def triple : Separate<["-"], "triple">, Alias<target>;
 
+def delayed_function_body_parsing :
+  Flag<["-"], "delayed-function-body-parsing">,
+  HelpText<"Delay function body parsing until the end of all files">;
```
This probably shouldn't be in this PR.

davidungar replied:
Yup. Maybe it snuck back in on a rebase. Deleted.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -223,52 +235,86 @@ class ArgsToFrontendInputsConverter {
+  /// Returns the output filenames on the command line or in the output
+  /// filelist. If there
+  /// were neither -o's nor an output filelist, returns an empty vector.
+  const std::vector<std::string> &getOutputFilenamesFromCommandLineOrFilelist();
```
Now that this has a simpler type, is there any reason not to use ArrayRef here?

davidungar replied:
D'Oh! Done.

jrose-apple commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -493,190 +612,208 @@ static bool ParseFrontendArgs(FrontendOptions &Opts, ArgList &Args,
+    return false;
+  }
+  // Only used for testing & when invoking frontend directly.
+  return deriveOutputFilenameForDirectory(outputFilename);
```
Yeah, this came out much cleaner!

davidungar replied:
:)

jrose-apple reviewed (commented):
This generally looks good, only a few remaining comments.

A meta-point: including the changes for the *previous* PR in this one made it harder to review again. I'm happy to review by either full diff or commit-by-commit, but including all the "have"/"has" changes in this diff made it harder to read in full, and separating it out into a different commit wouldn't have helped because the main contents of the PR are spread across multiple, non-self-contained commits.

Commit:
Review fixes:

Rename isThereAPrimaryInputWithAFilenameAt to isInputPrimary.
Remove spurious delayed_function_body_parsing.
Simply return type of getOutputFilenamesFromCommandLineOrFilelist.

davidungar commented:
@swift-ci please smoke test

davidungar commented:
Sorry about including the previous changes. I need to fix my workflow. Will discuss how.

davidungar commented:
@swift-ci please smoke test linux platform

davidungar commented:
@swift-ci please test os x platform
